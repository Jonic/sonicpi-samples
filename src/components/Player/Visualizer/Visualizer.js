import React, { Component } from 'react'
import PropTypes from 'prop-types'

class Visualizer extends Component {
  render() {
    return (
      <div className="c-Visualizer">
        {this.props.visData}
      </div>
    )
  }
}

Visualizer.propTypes = {
  visData:   PropTypes.object,
  isPlaying: PropTypes.bool.isRequired,
}

export default Visualizer
