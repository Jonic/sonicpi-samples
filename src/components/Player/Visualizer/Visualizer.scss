@import '../../../styles/component-helper';

.c-visualizer {}
