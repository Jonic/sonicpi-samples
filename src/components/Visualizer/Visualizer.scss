@import '../../styles/component-helper';

.c-visualizer {
  display: none;

  @media screen and (prefers-reduced-motion: no-preference) {
    display: block;
    left: 0;
    position: absolute;
    top: 0;
    z-index: 10;
  }
}
