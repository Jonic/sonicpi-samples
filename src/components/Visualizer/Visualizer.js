/* eslint-disable no-magic-numbers */
import './Visualizer.scss'

import React, { useContext, useEffect, useRef, useState } from 'react'

import AppContext from 'contexts/App'

const Visualizer = () => {
  const canvas = useRef(null)
  const [context, setContext] = useState(null)

  const [canvasHeight, setCanvasHeight] = useState(0)
  const [canvasWidth, setCanvasWidth] = useState(0)
  const { visData } = useContext(AppContext)

  useEffect(() => {
    setContext(canvas.current.getContext('2d'))
  }, [canvas])

  useEffect(() => {
    window.addEventListener('resize', updateWindowDimensions)
    updateWindowDimensions()
  })

  const updateWindowDimensions = () => {
    setCanvasHeight(window.innerHeight)
    setCanvasWidth(window.innerWidth)
  }

  useEffect(() => {
    canvas.current.width = canvasWidth

    let barWidth = (canvasWidth / visData.length) * 2
    let barHeight
    let colorOpacity
    let maxValue = 255
    let x = 0

    for (let value of visData) {
      barHeight = canvasHeight / (maxValue / value)
      colorOpacity = (1 - 100 / barHeight) / 1.5
      context.fillStyle = `rgba(255, 255, 255, ${colorOpacity})`
      context.fillRect(x, canvasHeight - barHeight, barWidth, barHeight)
      x += barWidth
    }
  }, [canvasHeight, canvasWidth, context, visData])

  return (
    <canvas
      className="c-visualizer"
      height={canvasHeight}
      ref={canvas}
      style={{
        height: canvasHeight,
        width: canvasWidth,
      }}
      width={canvasWidth}
    />
  )
}

export default Visualizer
