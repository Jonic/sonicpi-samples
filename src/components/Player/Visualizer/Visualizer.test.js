import React from 'react'
import { mount, shallow } from 'enzyme'
import Visualizer from './Visualizer'

let wrapper

describe('<Visualizer />', () => {
  beforeEach(() => {
    wrapper = shallow(
      <Visualizer />
    )
  })

  it('renders without crashing', () => {
    const mountedWrapper = mount(
      <Visualizer />
    )
    expect(mountedWrapper.find('.c-Visualizer')).toBePresent()
  })
})
