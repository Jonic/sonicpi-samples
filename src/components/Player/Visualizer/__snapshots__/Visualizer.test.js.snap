// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Visualizer /> Renders... matches its snapshot 1`] = `
<canvas
  className="c-visualizer"
  height={768}
  style={
    Object {
      "height": 768,
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": 1024,
      "zIndex": -1,
    }
  }
  width={1024}
/>
`;
